gateway-api: Fix GRPCRoute status not being displayed correctly

Archonr opened the pull request:
## Summary

This PR enables support for `GRPCRoute` attachment to `Gateway` listeners using the `HTTP` or `HTTPS` protocol in Cilium's Gateway API implementation.

## Description

Previously, Cilium did not properly support `GRPCRoute` resources — while traffic could be routed if configured manually, listeners on a `Gateway` would not register these routes as attached, and `kubectl describe gateway` would not show them under `Supported Kinds` or `Attached Routes`.

This change introduces the necessary wiring to support `GRPCRoute`:

- Registers `GRPCRoute` into the runtime scheme
- Updates the controller logic to support `GRPCRoute` as a valid kind for `HTTP` and `HTTPS` protocols
- Adds filtering and route attachment logic for `GRPCRoute`
- Ensures listener status is correctly populated with GRPCRoute data

## Problem

Even though `GRPCRoute` is a valid Gateway API object and often used with Envoy-based L7 load balancing, Cilium’s Gateway controller did not consider it for attachment or status reflection, causing:

- `Attached Routes: 0` when a `GRPCRoute` was attached
- `Supported Kinds` only showing `HTTPRoute`
- Inconsistent observability and diagnostics

## Solution

This PR includes:

- `gatewayv1.AddToScheme()` to explicitly register `GRPCRoute`
- Modifications in `setListenerStatus()` to:
  - Include GRPCRoutes in `supportedKinds`
  - Count attached GRPCRoutes
- Addition of `filterGRPCRoutesByListener()` to select matching GRPCRoutes for each listener
- Enhanced `getSupportedRouteKinds()` to return both `HTTPRoute` and `GRPCRoute` for HTTP/HTTPS listeners

## How to test

1. Deploy the patched operator image.
2. Create a `Gateway` with a listener of protocol `HTTP` or `HTTPS`.
3. Create a valid `GRPCRoute` that attaches to the Gateway.
4. Run:

   ```bash
   kubectl describe gateway <your-gateway-name>
   ```

5. Confirm:
   - `Attached Routes` is incremented
   - `Supported Kinds` includes `GRPCRoute`
   - gRPC traffic is routed correctly

## Notes

- This implementation is compatible with the current Gateway API spec
- Backward compatible with existing `HTTPRoute`, `TLSRoute`, etc.
- Further testing in CI environments may be warranted

## Related Issues

Fixes: https://github.com/cilium/cilium/issues/39021

Archonr committed:
gateway-api: support attaching GRPCRoute to HTTP/HTTPS listeners

This patch extends the Cilium Gateway API implementation to support GRPCRoute
attachment to HTTP/HTTPS listeners, following the spec behavior similar to Istio.

- Adds GRPCRoute to listener's attachedRoutes
- Allows AllowedRoutes to include GRPCRoute for HTTP/HTTPS
- Updates listener status logic to recognize GRPCRoutes

Fixes: #39021

Signed-off-by: Archonr <archonr94@gmail.com>
Signed-off-by: Vlad Korkach <v.korkach@playson.com>

sayboras commented on operator/pkg/gateway-api/cell.go:
```diff
@@ -324,6 +324,11 @@ func registerGatewayAPITypesToScheme(scheme *runtime.Scheme, optionalKinds []sch
+	if err := gatewayv1.AddToScheme(scheme); err != nil {
+		return fmt.Errorf("failed to register GRPCRoute: %w", err)
+	}
+
```
any reason to handle this one explicitly? Is it already done with other gatewayv1 resources?

Archonr replied:
You're right — calling `gatewayv1.AddToScheme(scheme)` explicitly is redundant here, since `gatewayv1.GroupVersion` is already included in the `addToSchema` map with `gatewayv1.AddToScheme`. That call ensures all types in the v1 API group, including GRPCRoute, are registered.

sayboras commented on operator/pkg/gateway-api/gateway_reconcile.go:
```diff
@@ -448,99 +467,47 @@ func (r *gatewayReconciler) setListenerStatus(ctx context.Context, gw *gatewayv1
-			})
-		}
 
-		if l.TLS != nil {
```
any reason to remove this TLS validation block ?

Archonr replied:
Thanks for pointing that out — that removal was unintentional. It must’ve happened while I was refactoring the `setListenerStatus` logic to support multiple route kinds (HTTP, gRPC). I’ll restore the TLS validation block and make sure it's preserved correctly alongside the new changes.

sayboras reviewed (commented):
Thanks for your contribution, it seems like the changes remove some of the existing logic as well

Archonr committed:
gateway-api: support GRPCRoute alongside HTTPRoute by default

By default, for HTTP and HTTPS protocols we consider both HTTPRoute and GRPCRoute as supported.
If AllowedRoutes.Kinds is explicitly defined, we filter the supported kinds down to those intersecting with what's explicitly allowed.
This aligns with the Gateway API spec and ensures conformance tests like GatewayWithAttachedRoutesWithPort8080 pass correctly.

Fixes: #39021

Signed-off-by: Archonr <archonr94@gmail.cim>
Signed-off-by: Vlad Korkach <v.korkach@playson.com>

Archonr committed:
gateway-api: align SupportedKinds handling with Gateway API spec

This patch refactors the logic for setting SupportedKinds on listeners
to better align with the Gateway API spec.

- For HTTP/HTTPS protocols, both HTTPRoute and GRPCRoute are returned by default.
- If AllowedRoutes.Kinds is set, only the intersection with supported kinds is allowed.
- Added strict validation to mark listeners invalid if explicitly allowed kinds are unsupported.
- Fixed group dereferencing and comparison issues between *Group and string.
- Ensures conformance test `GatewayWithAttachedRoutesWithPort8080` passes by always including HTTPRoute in supported kinds.

This improves conformance and prevents unexpected behavior when GRPCRoute is not installed.

Fixes: #39021

Signed-off-by: Archonr <archonr94@gmail.com>
Signed-off-by: Vlad Korkach <v.korkach@playson.com>

Archonr committed:
gateway-api: fix missing ObservedGeneration in listener conditions

Fixes conformance failure where Gateway listener conditions do not include
ObservedGeneration, causing stale condition errors during test validation.

This aligns with expectations from conformance test helper:
https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/utils/kubernetes/helpers.go#L115

Fixes: #39021

Signed-off-by: Archonr <archonr94@gmail.com>
Signed-off-by: Vlad Korkach <v.korkach@playson.com>

Archonr committed:
gateway-api: fix incorrect condition reason for ListenerConditionAccepted

This aligns with expectations from conformance test helper:
https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1/gateway_types.go#L1322

Fixes: #39021

Signed-off-by: Archonr <archonr94@gmail.com>
Signed-off-by: Vlad Korkach <v.korkach@playson.com>

Archonr committed:
gateway-api: fix listener conditions for invalid routes and resolved refs

- Ensure Programmed, Accepted, and ResolvedRefs conditions are set with correct ObservedGeneration and Reason for invalid RouteKinds and ReferenceGrant cases.
- Always set ResolvedRefs with Reason InvalidRouteKinds or RefNotPermitted depending on error type.

Signed-off-by: Archonr <archonr94@gmail.com>
Signed-off-by: Vlad Korkach <v.korkach@playson.com>

Archonr committed:
gateway-api: fix listener conditions for invalid routes and resolved

Signed-off-by: Archonr <archonr94@gmail.com>
Signed-off-by: Vlad Korkach <v.korkach@playson.com>

youngnick commented on operator/pkg/gateway-api/gateway_reconcile.go:
```diff
@@ -534,18 +541,35 @@ func (r *gatewayReconciler) setListenerStatus(ctx context.Context, gw *gatewayv1
-				Message:            "Invalid CertificateRef",
-				LastTransitionTime: metav1.Now(),
-			})
+			conds = append(conds,
+				gatewayListenerInvalidRouteKinds(gw, "Invalid Route Kinds"),
+				gatewayListenerAcceptedCondition(gw, false, "Listener not valid"),
+				metav1.Condition{
+					Type:               string(gatewayv1.ListenerConditionResolvedRefs),
+					Status:             metav1.ConditionFalse,
+					Reason:             "InvalidRouteKinds",
+					Message:            "Invalid Route Kinds",
+					ObservedGeneration: gw.GetGeneration(),
+					LastTransitionTime: metav1.Now(),
+				})
+		} else {
+			conds = append(conds,
+				gatewayListenerProgrammedCondition(gw, true, "Listener Programmed"),
+				gatewayListenerAcceptedCondition(gw, true, "Listener Accepted"),
+				metav1.Condition{
+					Type:               string(gatewayv1.ListenerConditionResolvedRefs),
+					Status:             metav1.ConditionTrue,
+					Reason:             string(gatewayv1.ListenerReasonResolvedRefs),
+					Message:            "Resolved Refs",
+					ObservedGeneration: gw.GetGeneration(),
+					LastTransitionTime: metav1.Now(),
+				})
```
Couple of things to fix here:

* all changes to the `conds` slice should use the `merge()` method, not `append`, as the `conds` slice is actually a `listTypeMap`, which works more like a map - there must only ever be one Condition of each Type in the array.
* The additions of the InvalidRouteKinds changes should happen in the earlier section, where we are actually checking the kinds, rather than here at the end. This `if !isValid` section should only have overarching changes, not specific ones.
* Also, having _some_ kinds not supported should not render the Listener as not `Accepted`, so we should only be setting the ResolvedRefs condition with the `InvalidRouteKinds` Condition in that case.
* Also, the `gatewayListenerInvalidRouteKinds()` function already adds the ResolvedRefs Condition, so you don't need to add it again.

youngnick commented on operator/pkg/gateway-api/gateway_reconcile.go:
```diff
@@ -452,46 +470,35 @@ func (r *gatewayReconciler) setListenerStatus(ctx context.Context, gw *gatewayv1
-					Group: g,
-					Kind:  k,
-				},
+			supportedKinds = allSupported
+			if len(supportedKinds) == 0 {
+				isValid = false
```
This branch is _very_ unclear. What it is actually checking is if the Protocol is not matched, so `allSupported` will be empty.

If we don't support any Route kinds for the Protocol, we should drop out straight away after the `allSupported` assignment.

Please remove the assignment and move the length check to be for `allSupported`, and place it just after the `allSupported` assignment.
